ibtk/include/ibtk/HierarchyAveragedDataManager.h

```diff
@@ -0,0 +1,251 @@
+     */
+    ~HierarchyAveragedDataManager() = default;
+
+    /*!
+     * Clear the snapshots taken with this class.
+     */
+    void clearSnapshots();
+
+    /*!
+     * Update the time averaged snapshot. The update is computed through the relation
+     * ``` u_avg = u_avg + 1/N * (u - u_avg)```
+     * in which N is the number of updates to the snapshot. If this the first update after clearSnapshots() is called,
+     * we simply copy data.
+     *
+     * This function returns true if the snapshot is at a steady state, or that 1/N*||u - u_avg|| < threshold. Note that
+     * a weight patch index should be supplied to accurately compute the norm.
+     *
+     * If specified in the constructor, this function also writes visualization files for both the mean field and the
+     * deviation.
+     */
+    inline bool updateTimeAveragedSnapshot(int u_idx,
+                                           double time,
+                                           SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                           const int wgt_idx = IBTK::invalid_index,
+                                           double tol = 1.0e-8)
+    {
+        return updateTimeAveragedSnapshot(u_idx, time, hierarchy, d_mean_refine_type, wgt_idx, tol);
+    }
+    bool updateTimeAveragedSnapshot(int u_idx,
+                                    double time,
+                                    SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                    const std::string& mean_refine_type,
+                                    const int wgt_idx = IBTK::invalid_index,
+                                    double tol = 1.0e-8);
+
+    /*!
+     * Return true if all the tracked mean fields are at a steady state.
+     */
+    inline bool isAtPeriodicSteadyState()
+    {
+        bool steady_state = true;
+        for (const auto& idx_steady_state : d_idx_steady_state_map)
+        {
+            steady_state = steady_state && idx_steady_state.second;
+        }
+        return steady_state;
+    }
+
+    /*!
+     * Return whether the point at the specified time is at a periodic steady state.
+     */
+    inline bool isAtPeriodicSteadyState(double time, const double tol)
+    {
+        time = getTimePt(time, tol);
+        return d_idx_steady_state_map.at(time);
+    }
+
+    /*!
+     * Return the time points this class is storing.
+     */
+    inline const std::set<double>& getSnapshotTimePts()
+    {
+        return d_snapshot_time_pts;
+    }
+
+    /*!
+     * Returns the exact data point at which points are being stored in the internal map. Note that if a time point is
+     * not found within the specified tolerance, an unrecoverable error occurs.
+     */
+    double getTimePt(double time, double tol);
+
+    /*!
+     * Get the SnapshotCache object
+     */
+    const std::unique_ptr<SnapshotCache>& getSnapshotCache()
```

Comment:
```suggestion
    const SnapshotCache& getSnapshotCache()
```
There's no reason to return a unique pointer by const reference - we should just return the object directly (by reference).